packages/flutter_tools/lib/src/template.dart

```diff
@@ -0,0 +1,132 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:path/path.dart' as path;
+import 'package:mustache4dart/mustache4dart.dart' as mustache;
+
+import 'artifacts.dart';
+import 'globals.dart';
+
+const String _kTemplateExtension = '.tmpl';
+const String _kCopyTemplateExtension = '.copy.tmpl';
+
+/// Expands templates in a directory to a destination. All files that must
+/// undergo template expansion should end with the '.tmpl' extension. All other
+/// files are ignored. In case the contents of entire directories must be copied
+/// as is, the directory itself can end with '.tmpl' extension. Files within
+/// such a directory may also contain the '.tmpl' extension and will be
+/// considered for expansion. In case certain files need to be copied but
+/// without template expansion (images, data files, etc.), the '.copy.tmpl'
+/// extension may be used.
+///
+/// Files in the destination will not contain either the '.tmpl' or '.copy.tmpl'
+/// extensions.
+class Template {
+  factory Template.fromName(String name) {
```

Comment:
I tend to avoid factory ctos unless I'm using it to return one of a few different subclasses. You might want to use a static method here, or have an `_init` method that both ctors called into.